src/Components/Components/src/Rendering/Renderer.cs

```diff
@@ -452,18 +452,95 @@ private void ProcessRenderQueue()
             }
         }
 
-        private Task InvokeRenderCompletedCalls(ArrayRange<RenderTreeDiff> updatedComponents)
+        private Task InvokeRenderCompletedCalls(ArrayRange<RenderTreeDiff> updatedComponents, Task updateDisplayTask)
         {
+            if (updateDisplayTask.IsCanceled)
+            {
+                // The display update was cancelled (maybe due to a timeout on the components server-side case or due
+                // to the renderer being disposed)
+                return Task.CompletedTask;
+            }
+            if (updateDisplayTask.IsFaulted)
+            {
+                // The display update failed so we don't care any more about running on render completed
+                // fallbacks as the entire rendering process is going to be torn down.
+                HandleException(updateDisplayTask.Exception);
+                return Task.CompletedTask;
+            }
+
+            if (!updateDisplayTask.IsCompleted)
+            {
+                var updatedComponentsId = new int[updatedComponents.Count];
+                var updatedComponentsArray = updatedComponents.Array;
+                for (int i = 0; i < updatedComponentsId.Length; i++)
+                {
+                    updatedComponentsId[i] = updatedComponentsArray[i].ComponentId;
+                }
+
+                return InvokeAfterUpdateDisplayCompleted(updateDisplayTask, updatedComponentsId);
+            }
+
             List<Task> batch = null;
             var array = updatedComponents.Array;
             for (var i = 0; i < updatedComponents.Count; i++)
             {
                 var componentState = GetOptionalComponentState(array[i].ComponentId);
                 if (componentState != null)
+                {
+                    ProcessComponent(componentState, batch);
+                }
+            }
+
+            return batch != null ?
+                Task.WhenAll(batch) :
+                Task.CompletedTask;
+
+            async Task InvokeAfterUpdateDisplayCompleted(Task updateDisplayTask, int[] updatedComponents)
+            {
+                try
+                {
+                    await updateDisplayTask;
+                }
+                catch when (updateDisplayTask.IsCanceled)
+                {
+                    return;
+                }
+                catch when (updateDisplayTask.IsFaulted)
+                {
+                    HandleException(updateDisplayTask.Exception);
+                    return;
+                }
+
+                List<Task> batch = null;
```

Comment:
where is batch set to non-null?

Reply:
You right. Seems like we don't have a test for this.

Reply:
I added a couple of tests to cover this.